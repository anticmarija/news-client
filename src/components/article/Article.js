import React from "react";

const Article = () => {
  //   const { title, urlToImage, description } = article;

  return (
    <div>
      article
      {/* <h1>{title}</h1>
      <img alt={title} src={urlToImage} />
      <p>{description}</p> */}
    </div>
  );
};

export default Article;
