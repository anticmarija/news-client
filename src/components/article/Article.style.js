import styled from "styled-components/macro";

export const ArticleWrapperStyle = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
`;

export const ArticleImageStyle = styled.img`
  width: 50%;
  height: auto;
  margin-top: 50px;
`;
