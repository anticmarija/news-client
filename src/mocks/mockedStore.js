import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
export const initialState = {
  topNews: {},
  countries: {
    supportedCountries: [{ countryId: "gb", countryName: "Gb" }],
  },
  categories: {},
  article: {
    activeArticle: {
      title: "active",
      urlToImage:
        "https://i.dailymail.co.uk/1s/2020/06/17/16/29719992-0-image-a-23_1592408042536.jpg",
      description: "desc",
    },
  },
};

export const mockedStore = (state = initialState) => {
  const mockStore = configureStore(middlewares);
  return mockStore(state);
};
