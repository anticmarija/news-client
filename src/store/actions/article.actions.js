export const SET_ACTIVE_ARTICLE = "SET_ACTIVE_ARTICLE";

/**
 * Action creators
 */
export const setActiveArticle = (article) => {
  return { type: SET_ACTIVE_ARTICLE, payload: article };
};
