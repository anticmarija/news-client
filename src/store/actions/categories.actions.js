export const SET_CATEGORIES = "SET_CATEGORIES";
/**
 * Action creators
 */
export const setCategories = (categories) => {
  return { type: SET_CATEGORIES, payload: categories };
};
