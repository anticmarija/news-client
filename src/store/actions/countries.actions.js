export const SET_COUNTRIES = "SET_COUNTRIES";
export const SET_ACTIVE_COUNTRY = "SET_ACTIVE_COUNTRY";
/**
 * Action creators
 */
export const setCountries = (countries) => {
  return { type: SET_COUNTRIES, payload: countries };
};

export const setActiveCountry = (country) => {
  return { type: SET_ACTIVE_COUNTRY, payload: country };
};
