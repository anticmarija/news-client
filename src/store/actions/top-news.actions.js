import config from "../../config";

export const FETCH_ARTICLES_START = "FETCH_ARTICLES_START";
export const FETCH_ARTICLES_SUCCESS = "FETCH_ARTICLES_SUCCESS";
export const FETCH_ARTICLES_FAIL = "FETCH_ARTICLES_FAIL";

/**
 * Action creators
 */
export const fetchArticlesStart = () => {
  return { type: FETCH_ARTICLES_START };
};

export const fetchArticlesSuccess = (articles) => {
  return { type: FETCH_ARTICLES_SUCCESS, payload: articles };
};

export const fetchArticlesFail = (error) => {
  return { type: FETCH_ARTICLES_FAIL, payload: error };
};

/**
 * Fetches ARTICLES and dispatches success and error actions
 */
export const fetchArticles = (countryId) => {
  return async (dispatch) => {
    try {
      dispatch(fetchArticlesStart());

      const response = await fetch(
        `${config.getApiBaseUrl()}/top-headlines?country=${countryId}&apiKey=${config.getApiKey()}`
      );

      const { articles } = await response.json();

      dispatch(fetchArticlesSuccess(articles));
    } catch (err) {
      dispatch(fetchArticlesFail(err));
    }
  };
};
