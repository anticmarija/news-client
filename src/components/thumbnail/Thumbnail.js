import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { useSelector, useDispatch } from "react-redux";
import {
  ThumbnailWrapperStyle,
  ThumbnailImgStyle,
  ThumbnailTitleStyle,
  ThumbnailLinkStyle,
  ThumbnailDescStyle,
} from "./Thumbnail.style";
import { setActiveArticle } from "../../store/actions/article.actions";
import { disableChangingCountries } from "../../store/actions/countries.actions";
import { useTranslation } from "react-i18next";

const Thumbnail = ({ article }) => {
  const { t } = useTranslation();

  const { activeCountry } = useSelector((state) => state.countries);
  const { title, description, urlToImage } = article;

  const dispatch = useDispatch();

  return (
    <ThumbnailWrapperStyle>
      <ThumbnailTitleStyle>{title}</ThumbnailTitleStyle>
      <ThumbnailImgStyle imageUrl={urlToImage} />
      <ThumbnailDescStyle>
        <div>{description}</div>
        <ThumbnailLinkStyle>
          <Link
            data-testid="more-link"
            onClick={() => {
              dispatch(setActiveArticle(article));
              dispatch(disableChangingCountries());
            }}
            to={`/country/${activeCountry}/articles/${encodeURIComponent(
              title.toLowerCase().split(" ").join("-")
            )}`}
          >
            {t("thumbnail.readMore")} &gt;
          </Link>
        </ThumbnailLinkStyle>
      </ThumbnailDescStyle>
    </ThumbnailWrapperStyle>
  );
};

Thumbnail.propTypes = {
  article: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    urlToImage: PropTypes.string,
  }),
};

export default Thumbnail;
