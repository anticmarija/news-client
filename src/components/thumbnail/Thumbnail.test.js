import React from "react";
import Thumbnail from "./Thumbnail";
import { mockedStore } from "../../mocks/mockedStore";
import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import * as actionsCountries from "../../store/actions/countries.actions";
import * as actionsArticles from "../../store/actions/article.actions";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (s) => s }),
}));

const store = mockedStore();

describe("Test Thumbnail component", () => {
  it("should render correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Thumbnail
            article={{
              title: "title",
              description: "desc",
              urlToImage: "some-url",
            }}
          />
        </MemoryRouter>
      </Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should disable changing countries correctly", () => {
    const { queryByTestId } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Thumbnail
            article={{
              title: "title",
              description: "desc",
              urlToImage: "some-url",
            }}
          />
        </MemoryRouter>
      </Provider>
    );

    const countryLink = queryByTestId("more-link");

    const actionArticleSpy = jest.spyOn(actionsArticles, "setActiveArticle");
    const actionCountriesSpy = jest.spyOn(
      actionsCountries,
      "disableChangingCountries"
    );

    fireEvent.click(countryLink);

    expect(actionArticleSpy).toBeCalled();
    expect(actionCountriesSpy).toBeCalled();
  });
});
