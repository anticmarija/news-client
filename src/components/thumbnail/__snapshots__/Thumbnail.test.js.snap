// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Thumbnail component should render correctly 1`] = `
<div
  class="Thumbnailstyle__ThumbnailWrapperStyle-sc-123obdz-0 lhHvSI"
>
  <h2
    class="Thumbnailstyle__ThumbnailTitleStyle-sc-123obdz-3 dSnuxn"
  >
    title
  </h2>
  <div
    class="Thumbnailstyle__ThumbnailImgStyle-sc-123obdz-1 jrgXyZ"
  />
  <div
    class="Thumbnailstyle__ThumbnailDescStyle-sc-123obdz-2 dPDhAr"
  >
    <div>
      desc
    </div>
    <span
      class="Thumbnailstyle__ThumbnailLinkStyle-sc-123obdz-4 VvDHO"
    >
      <a
        data-testid="more-link"
        href="/country/gb/articles/title"
      >
        thumbnail.readMore
         &gt;
      </a>
    </span>
  </div>
</div>
`;
